tools/fuzzer/afl-main.c

```diff
@@ -0,0 +1,498 @@
+		switch (opt) {
+		case 'i':
+			ipcmsg_file = optarg;
+			break;
+		case 'p':
+			platform_name = optarg;
+			break;
+		case 'w':
+			qemu_wait = parse_qemu_wait(optarg);
+			break;
+		case 'h':
+			usage(argv[0]);
+			exit(0);
+			break;
+		default:
+			usage(argv[0]);
+			exit(EXIT_FAILURE);
+		}
+	}
+
+	/* initialise emulated target device */
+	if (!platform_name) {
+		fprintf(stderr, "error: no target platform specified\n");
+		usage(argv[0]);
+		exit(EXIT_FAILURE);
+	}
+
+	/* initialise emulated target device */
+	if (!ipcmsg_file) {
+		fprintf(stderr, "error: no ipc message file specified\n");
+		usage(argv[0]);
+		exit(EXIT_FAILURE);
+	}
+
+	/* find platform */
+	for (i = 0; i < ARRAY_SIZE(platform); i++) {
+		if (!strcmp(platform[i]->name, platform_name))
+			break;
+	}
+
+	/* no platform found */
+	if (i >= ARRAY_SIZE(platform)) {
+		fprintf(stderr, "error: platform %s not supported\n",
+			platform_name);
+		usage(argv[0]);
+		exit(EXIT_FAILURE);
+	}
+
+	pid_t fid = fork();
+
+	if (!fid) {
+
+		/*
+		 * Hardcode qemu-system-xtensa path for byt platform.
+		 * TODO: path and options should be different for different platforms
+		 */
+		static char *newargv[] = {"/home/sof/work/qemu/build/xtensa-softmmu/qemu-system-xtensa",
+				    "-cpu", "baytrail", "-M", "adsp_byt", "-nographic", "-kernel",
+				    "/home/sof/work/sof.git/build_byt_gcc/src/arch/xtensa/sof-byt.ri",
+				    NULL};
+		static char *newenviron[] = { NULL };
+
+		fprintf(stdout, "Executing qemu xtensa\n");
+		execve(newargv[0], newargv, newenviron);
+
+		/* execve() returns only on error */
+		perror("execve");
+		exit(EXIT_FAILURE);
+	}
+
+	/* TODO: Need to make sure that child is executed first. */
+	fprintf(stdout, "Waiting for %u secs for qemu DSP to start\n", qemu_wait);
+	sleep(qemu_wait);
+
+	/* init plaform */
+	fprintf(stdout, "initialising platform %s\n", platform[i]->name);
+	ret = platform[i]->init(&fuzzer, platform[i]);
+	if (ret == ETIMEDOUT) {
+		fprintf(stderr, "error: platform %s failed to initialise\n",
+			platform_name);
+		exit(EXIT_FAILURE);
+	}
+
+	fprintf(stdout, "FW boot complete\n");
+
+	/* allocate max ipc size bytes for the msg and reply */
+	fuzzer.msg.msg_data = malloc(SOF_IPC_MSG_MAX_SIZE);
+	fuzzer.msg.reply_data = malloc(SOF_IPC_MSG_MAX_SIZE);
+
+	/* load ipc message */
+	ret = parse_ipcmsg(&fuzzer, ipcmsg_file);
+	if (ret < 0)
+		exit(EXIT_FAILURE);
+
+	/* send ipc message */
+	ret = fuzzer_send_msg(&fuzzer);
+
+	/* all done - now free platform */
+	platform[i]->free(&fuzzer);
+
+	/* As it's not guaranteed that library functions set it to zero on success */
+	errno = 0;
+
+	/* kill the DSP process created above */
+	if (kill(fid, SIGKILL))
+		fprintf(stderr, "killing child process failed\n");
+
+	/* ipc message reply timeout */
+	if (ret == -EINVAL) {
+		fprintf(stderr, "error: failed to receive reply from DSP\n");
+
+		/* raise SIGABRT so that fuzzer sees this as a crash */
+		abort();
+	}
+
+	return -errno;
```

Comment:
isn't errno hardcoded to 0 here? we should just return 0

Reply:
There is a possibility that kill() system call may fail above. If it fails errno is set appropriately. This explains the use of errno. If it doesn't fail we want to return 0. so the errno was overwritten to 0 above.